scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetListFormatRule.scala

```diff
@@ -0,0 +1,262 @@
+            typ = elementType,
+            name = elementType.getName,
+            isElementRequired = sourceRule.isElementRequired(repeatedSourceType),
+            originalType = sourceRule.elementOriginalType(repeatedSourceType)
+          )
+        }
+      }
+      case _ => repeatedSourceType
+    }
+  }
+
+  def isGroupList(projection: Type): Boolean = {
+    if (projection.isPrimitive) {
+      false
+    } else {
+      val groupProjection = projection.asGroupType
+      groupProjection.getOriginalType == OriginalType.LIST &&
+        groupProjection.getFieldCount == 1 &&
+        groupProjection.getFields.get(0).isRepetition(Type.Repetition.REPEATED)
+    }
+
+  }
+
+  def findFirstListRule(repeatedType: Type,
+                        sourceOrTarget: SourceOrTarget): Option[ParquetListFormatRule] = {
+    val ruleFound = sourceOrTarget.rules.find(rule => rule.check(repeatedType))
+    if (ruleFound.isEmpty) LOGGER.warn(s"Unable to find matching rule for ${sourceOrTarget.name} schema:\n$repeatedType")
+    ruleFound
+  }
+
+  def wrapElementAsRepeatedType(rule: ParquetListFormatRule, repeatedType: Type, elementType: Type): Type = rule.createCompliantRepeatedType(
+    elementType,
+    rule.elementName(repeatedType),
+    // if repeated or required, it is required
+    !elementType.isRepetition(Type.Repetition.OPTIONAL),
+    elementType.getOriginalType)
+}
+
+/**
+ * Helper to specify supported source/target conversion.
+ */
+private[scrooge] sealed trait SourceOrTarget {
+  def rules: Seq[ParquetListFormatRule]
+  def name: String
+}
+
+private[scrooge] object Source extends SourceOrTarget {
+  override val rules: Seq[ParquetListFormatRule] = Seq(
+    PrimitiveElementRule, PrimitiveArrayRule,
+    GroupElementRule, GroupArrayRule,
+    TupleRule, StandardRule
+  )
+
+  override def name: String = "source"
+}
+
+private[scrooge] object Target extends SourceOrTarget {
+  override def rules: Seq[ParquetListFormatRule] = Source.rules.filterNot(_ == TupleRule)
+
+  override def name: String = "target"
+}
+
+/**
+ * Rule allowing conversion from one format to other format by
+ * 1) detect which format is the repeated list type.
+ * 2) decompose the repeated type into element and other info.
+ * 3) construct compliant repeated type from the given element and other info.
+ * For example,
+ * if source repeated type matches Rule 1, and target type matches Rule 2.
+ * Rule 1 will decompose the source type, and
+ * Rule 2 will take that information to construct repeated element in Rule 2 format.
+ */
+private[scrooge] sealed trait ParquetListFormatRule {
+  def elementType(repeatedType: Type): Type
+
+  def elementName(repeatedType: Type): String = this.elementType(repeatedType).getName
+
+  def elementOriginalType(repeatedType: Type): OriginalType = this.elementType(repeatedType).getOriginalType
+
+  private[scrooge] def isElementRequired(repeatedType: Type): Boolean
+
+  private[scrooge] def check(typ: Type): Boolean
```

Comment:
From the name, it's a bit non-obvious what this method does. It this actually something like `appliesToType(typ: Type)` which returns `true` if the type is a list of this format and `false` otherwise?

Reply:
+1 on naming this something more descriptive

Reply:
Updated as suggested